File: servers/python-minimal/apron_server.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.10"
# dependencies = ["websockets>=14,<17"]
# ///
"""Apron Chat v3 for trusted, compliant clients.

Run: uv run apron_server.py [--host 0.0.0.0] [--port 8765]
Or:  python -m pip install 'websockets>=14,<17'; python apron_server.py

One room, guest identities, names, replies, ext pass-through, and the latest
1,000 log records of history. No edits, room changes, reactions, persistence,
or retry deduplication. Reconnects get new IDs. Input shapes/types are assumed
valid; malformed input may close a connection instead of returning protocol
errors. No outgoing buffer limits.
"""

import argparse
import asyncio
import json
import time
from collections import deque
from uuid import uuid4

from websockets.asyncio.server import broadcast, serve
from websockets.exceptions import ConnectionClosed


GREETING = {"protocol": 3, "name": "apron-python/3", "auth": ["guest"],
            "caps": ["history"]}


class Error(Exception):
    def __init__(self, code, message):
        self.payload = {"code": code, "message": message}


def require(condition, message="Invalid parameters", code=-32602):
    if not condition:
        raise Error(code, message)


def send(peers, **frame):
    # Synchronous writes preserve room order without awaiting slow clients.
    broadcast(peers, json.dumps(frame, ensure_ascii=False))


class ApronServer:
    def __init__(self, history_size=1000):
        self.clients = {}  # Only authenticated connections receive broadcasts.
        self.log = deque(maxlen=history_size)  # Room and message records.
        self.last_id = 0
        self.room = {"room_id": "general", "log_id": self.next_id(), "title": "General"}
        self.log.append(self.room)

    def next_id(self):
        self.last_id = max(time.time_ns() // 1_000_000, self.last_id + 1)
        return str(self.last_id)

    def availability(self):
        return {"latest_log_id": str(self.last_id), "history_log_id": self.log[0]["log_id"]}

    def history(self, p):
        limit = p.get("limit", 50)
        require(type(limit) is int and limit > 0, "Invalid limit")
        require(all(str(p.get(k, "0")).isdigit() for k in ("after", "before")), "Invalid bounds")
        limit = min(limit, 200)
        after, before = int(p.get("after", "0")), int(p.get("before", self.last_id))
        matches = [r for r in self.log if after <= int(r["log_id"]) <= before]
        page = matches[:limit] if "after" in p else matches[-limit:]
        result = {"rooms": [r for r in page if "message_id" not in r],
                  "entries": [r for r in page if "message_id" in r],
                  "more": len(matches) > limit, **self.availability()}
        if page:
            result.update(first_id=page[0]["log_id"], last_id=page[-1]["log_id"])
        return result

    def create_message(self, you, p):
        require(isinstance(p.get("body"), dict), "Missing body")
        require(not p.get("deleted"), "Cannot create a deleted message")
        reply_to = p.get("reply_to")
        # The new ID is minted below, so a known target can never be the message itself.
        if "reply_to" in p:
            target = reply_to.get("message_id") if isinstance(reply_to, dict) else None
            require(isinstance(target, str) and any(r.get("message_id") == target for r in self.log),
                    "Unknown reply target")
        message_id = self.next_id()
        message = {"message_id": message_id, "log_id": message_id, "room_id": "general",
                   "from": dict(you), "body": {"format": "plain", **p["body"]}}
        if reply_to:
            message["reply_to"] = {"message_id": reply_to["message_id"]}
        if p.get("ext"):
            message["ext"] = p["ext"]
        self.log.append(message)
        send(self.clients, method="message", params=message)
        return {"message_id": message_id}

    def dispatch(self, ws, method, p):
        # No awaits: mutation, history selection, and live delivery are atomic
        # with respect to other connections on this event loop.
        if method == "auth" and ws not in self.clients:  # Any scheme is accepted.
            self.clients[ws] = {"user_id": "guest_" + uuid4().hex}

        require(ws in self.clients, "Authenticate first", -32001)
        if method in ("auth", "name"):
            if isinstance(p.get("name"), str):
                self.clients[ws] = {**self.clients[ws], "name": p["name"]}
            return {"you": self.clients[ws]}

        require(method in ("message", "history"), "Unsupported method", -32601)
        require(method != "message" or "message_id" not in p, "Editing is unsupported", -32601)
        require(p.get("room_id") == "general", "Unknown room")
        return self.history(p) if method == "history" else self.create_message(self.clients[ws], p)

    def receive(self, ws, raw):
        frame = json.loads(raw)
        try:
            result = self.dispatch(ws, frame["method"], frame.get("params", {}))
            if "id" in frame:
                send([ws], id=frame["id"], result=result)
            if frame["method"] == "auth":
                send([ws], method="room", params={**self.room, **self.availability()})
        except Error as error:
            if "id" in frame:
                send([ws], id=frame["id"], error=error.payload)

    async def handle(self, ws):
        send([ws], method="server", params=GREETING)
        try:
            async for raw in ws:
                self.receive(ws, raw)
        except ConnectionClosed:
            pass
        finally:
            self.clients.pop(ws, None)


async def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, default=8765)
    args = parser.parse_args()
    server = ApronServer()
    async with serve(server.handle, args.host, args.port, max_size=256 * 1024):
        print(f"Apron Chat listening on ws://{args.host}:{args.port}", flush=True)
        await asyncio.Future()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        pass

File: servers/js-minimal/apron-server.js
// Apron Chat v3 for trusted, compliant clients. Requires Bun; no dependencies.
// Run: bun apron-server.js
// LAN: HOST=0.0.0.0 PORT=8765 bun apron-server.js
// One room, guest identities, names, replies, ext pass-through, and the latest
// 1,000 log records of RAM history. Restart clears history; reconnect assigns a
// new identity. No retry deduplication, edits, room changes, reactions, uploads,
// credentials, or rate limits. Clients must send valid protocol frames;
// malformed input may close the connection.

const greeting = { protocol: 3, name: "apron-bun/3", auth: ["guest"], caps: ["history"] };
const log = []; // Room and message records, ascending by log_id.
let lastLogId = 0;

function nextLogId() {
  lastLogId = Math.max(Date.now(), lastLogId + 1);
  return String(lastLogId);
}

function append(record) {
  log.push(record);
  if (log.length > 1000) log.shift();
  return record;
}

const room = append({ room_id: "general", log_id: nextLogId(), title: "General" });

function send(ws, frame) {
  ws.send(JSON.stringify(frame));
}

function check(condition, message, code = -32602) {
  if (!condition) throw { code, message };
}

function availability() {
  return { latest_log_id: String(lastLogId), history_log_id: log[0].log_id };
}

function readHistory(params) {
  const limit = Math.min(params.limit ?? 50, 200);
  check(Number.isInteger(limit) && limit > 0, "Invalid limit");
  check(["after", "before"].every(key => /^\d+$/.test(params[key] ?? "0")), "Invalid bounds");
  const after = Number(params.after ?? 0);
  const before = Number(params.before ?? lastLogId);
  const matches = log.filter(record => {
    const logId = Number(record.log_id);
    return logId >= after && logId <= before;
  });
  const slice = "after" in params ? matches.slice(0, limit) : matches.slice(-limit);
  const result = {
    rooms: slice.filter(record => !record.message_id),
    entries: slice.filter(record => record.message_id),
    more: matches.length > limit,
    ...availability(),
  };
  if (slice.length) {
    result.first_id = slice[0].log_id;
    result.last_id = slice.at(-1).log_id;
  }
  return result;
}

function createMessage(you, params) {
  check(params.body && typeof params.body === "object", "Missing body");
  check(!params.deleted, "Cannot create a deleted message");
  const replyTo = params.reply_to?.message_id;
  // The new ID is minted below, so a known target can never be the message itself.
  if ("reply_to" in params) {
    check(typeof replyTo === "string" && log.some(record => record.message_id === replyTo), "Unknown reply target");
  }

  const id = nextLogId();
  const message = {
    message_id: id, log_id: id, room_id: "general", from: { ...you },
    body: { format: "plain", ...params.body },
  };
  if ("reply_to" in params) message.reply_to = { message_id: replyTo };
  if (params.ext) message.ext = params.ext;
  append(message);
  // Server.publish includes the sender; ws.publish would exclude it.
  server.publish("general", JSON.stringify({ method: "message", params: message }));
  return { message_id: id };
}

// Keep this synchronous: history reads, commits, and broadcasts must stay ordered.
function dispatch(ws, method, params) {
  if (method === "auth") ws.data.you ??= { user_id: "guest_" + crypto.randomUUID() }; // Any scheme.
  check(ws.data.you, "Authenticate first", -32001);
  if (method === "auth" || method === "name") {
    if (typeof params.name === "string") ws.data.you = { ...ws.data.you, name: params.name };
    return { you: ws.data.you };
  }

  check(method === "message" || method === "history", "Unsupported method", -32601);
  check(method !== "message" || !("message_id" in params), "Editing is unsupported", -32601);
  check(params.room_id === "general", "Unknown room");
  return method === "history" ? readHistory(params) : createMessage(ws.data.you, params);
}

const server = Bun.serve({
  hostname: process.env.HOST ?? "127.0.0.1",
  port: Number(process.env.PORT ?? 8765),
  fetch(req, server) {
    if (server.upgrade(req, { data: {} })) return;
    return new Response("WebSocket required", { status: 426 });
  },
  websocket: {
    maxPayloadLength: 256 * 1024,
    closeOnBackpressureLimit: true, // Disconnect slow clients instead of dropping live records.
    open(ws) {
      send(ws, { method: "server", params: greeting });
    },
    message(ws, raw) {
      let frame;
      try {
        frame = JSON.parse(raw);
        if (!frame || typeof frame !== "object" || Array.isArray(frame)) return ws.close(1003, "Invalid frame");
        const result = dispatch(ws, frame.method, frame.params ?? {});
        if ("id" in frame) send(ws, { id: frame.id, result });
        if (frame.method === "auth") {
          send(ws, { method: "room", params: { ...room, ...availability() } });
          ws.subscribe("general");
        }
      } catch (error) {
        if (!error.code) return ws.close(1003, "Invalid frame");
        if (frame?.id !== undefined) send(ws, { id: frame.id, error });
      }
    },
  },
});
console.log(`Apron Chat listening on ws://${server.hostname}:${server.port}`);
